### PR TITLE
[DS-1341] Contract Search Filters

### PR DESCRIPTION
**Pull request by ejgullo**
- [x] Code Reviewed
- [x] API Merged
- [x] PR #397  is merged

- Adding filters on search page for Pricing Type, Set Aside, and Extent Competed.  These filters only apply to contracts, so no financial assistance awards should display when one is selected.
- Show more/less functionality is in place
- Selected options have tags in the top bar that can be removed
- jest tests written for new reducer actions + container

Reviewed by @kevinli-work:
- [x] redux/reducers/search/searchFiltersReducer
- [x] redux/reducers/search/filters/contractFilterFunctions
- [x] redux/actions/search/searchFilterActions
- [x] models/search/SearchAwardsOperation
- [x] models/search/SearchOperation
- [x] models/search/SearchTransactionOperation
- [x] models/search/queryBuilders/ContractFilterQuery
- [x] models/search/queryBuilders/OtherFiltersQuery
- [x] dataMapping/search/contractFields
- [x] dataMapping/search/filterFields
- [x] helpers/sidebarHelper
- [x] containers/search/filters/ExtentCompetedContainer
- [x] containers/search/filters/PricingTypeContainer
- [x] containers/search/filters/SetAsideContainer
- [x] containers/search/topFilterBar/TopFilterBarContainer
- [x] components/search/filters/cfda/CFDASearch
- [x] components/search/filters/contractFilters/ContractFilter
- [x] components/search/filters/naics/NAICSSearch
- [x] components/search/filters/otherFilters/OtherFilters
- [x] components/search/filters/psc/PSCSearch
- [x] components/search/SearchSidebar
- [x] components/search/topFilterBar/filterGroups/ExtentCompetedFilterGroup
- [x] components/search/topFilterBar/filterGroups/PricingTypeFilterGroup
- [x] components/search/topFilterBar/filterGroups/SetAsideFilterGroup
- [x] components/search/topFilterBar/filterGroups/TopFilterGroupGenerator
- [x] components/sharedComponents/checkbox/SingleCheckboxType

### CODE REVIEW

**Comment by geoffgeiling** [comment]
All filters should show 5 by default instead of 6

screen shot 2017-06-28 at 11 33 15 am

**Comment by geoffgeiling** [comment]
All filter options should be in descending sort order
screen shot 2017-06-28 at 11 33 15 am

**Comment by geoffgeiling** [comment]
I recognize this isn't currently in the mock (and will have it changed) but after talking to Ross, we want to change the title of "Set-Aside Type" to "Type of Set-Aside"

**Comment by geoffgeiling** [comment]
I believe Extent Competed should have 9 options? So show first 5 be default and then "show 4 more"
screen shot 2017-06-28 at 3 48 45 pm

**Comment by kevinli-work** [comment]
It seems that most of these new filters do not work with the `/tas/categories/total` and `/transactions/total/` API endpoints. Please work with the backend team to validate what the fields should be.

File: tests/redux/reducers/search/searchFiltersReducer-test.js [hunk 1]
```diff
@@ -630,6 +629,39 @@ describe('searchFiltersReducer', () => {
         });
     });
 
+    describe('UPDATE_PRICING_TYPE', () => {
```
Comment (kevinli-work) [inline tests/redux/reducers/search/searchFiltersReducer-test.js]: The `contractFilterFunctions` add additional functionality besides simply creating a Set from a single value. Can you expand your tests to cover what happens if a value already exists (it should delete) or you try to add multiple values, some of which may already exist (it should append and/or delete)?

File: src/js/components/search/filters/contractFilters/ContractFilter.jsx [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    shown: defaultShown,
+    shownType: 'more'
+};
+
+export default class ContractFilter extends React.Component {
+
+    constructor(props) {
+        super(props);
+
+        this.state = defaultState;
```
Comment (kevinli-work) [inline src/js/components/search/filters/contractFilters/ContractFilter.jsx]: Why are you doing it this way?

File: src/js/components/search/filters/contractFilters/ContractFilter.jsx [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            updatedState = {
+                shown: Object.keys(contractFilters).length,
+                shownType: 'fewer'
+            };
+        }
+
+        this.setState(updatedState);
+    }
+
+    toggleValue(value) {
```
Comment (kevinli-work) [inline src/js/components/search/filters/contractFilters/ContractFilter.jsx]: Do we need this function? Can we just pass the prop straight down to `toggleCheckboxType`?

File: src/js/components/search/filters/contractFilters/ContractFilter.jsx [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+export default class ContractFilter extends React.Component {
+
+    constructor(props) {
+        super(props);
+
+        this.state = defaultState;
+
+        // Bind functions
+        this.toggleValue = this.toggleValue.bind(this);
+        this.generateContractFilters = this.generateContractFilters.bind(this);
```
Comment (kevinli-work) [inline src/js/components/search/filters/contractFilters/ContractFilter.jsx]: You don't need to bind this function if you're not passing it out of the component or calling it from a different context.

File: src/js/components/search/filters/contractFilters/ContractFilter.jsx [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    toggleValue(value) {
+        this.props.toggleFilter(value);
+    }
+
+    generateContractFilters(filters) {
+        const contractFilters = [];
+
+        if (contractFilters.length < this.state.shown) {
```
Comment (kevinli-work) [inline src/js/components/search/filters/contractFilters/ContractFilter.jsx]: Should this be `<=`?

File: src/js/components/search/filters/contractFilters/ContractFilter.jsx [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+            Object.keys(filters).forEach((key) => {
+                if (contractFilters.length <= this.state.shown
```
Comment (kevinli-work) [inline src/js/components/search/filters/contractFilters/ContractFilter.jsx]: Is `contractFilters.length <= this.state.shown` still necessary at this point? Won't the condition have to be true to enter the parent `if` statement?

File: src/_scss/pages/homepage/Untitled [hunk 7]
```diff
@@ -0,0 +1,50 @@
+/**
```
Comment (ejgullo) [inline src/_scss/pages/homepage/Untitled]: Whoops, should not have made it in.  Will delete.

File: src/js/components/search/filters/contractFilters/ContractFilter.jsx (hunk 2 above)
Comment (ejgullo) [inline src/js/components/search/filters/contractFilters/ContractFilter.jsx]: It's the way it's handled in the file I used as a guide - I can move the default values down if that's the preference.

File: src/js/components/search/filters/contractFilters/ContractFilter.jsx (hunk 3 above)
Comment (ejgullo) [inline src/js/components/search/filters/contractFilters/ContractFilter.jsx]: I don't believe so.  Removing.

File: src/js/models/search/SearchOperation.js [hunk 8]
```diff
@@ -152,6 +161,21 @@ class SearchOperation {
             }
         }
 
+        // Add Pricing Type Queries
+        if (this.pricingType.length > 0) {
```
Comment (ejgullo) [inline src/js/models/search/SearchOperation.js]: Adding searchContext to support the above issue w/ filters not hitting correct fields on various endpoints.

**Comment by ejgullo** [comment]
Ready for review, but should not be merged until v2 endpoints are complete.

File: src/js/dataMapping/search/filterFields.js [hunk 9]
```diff
@@ -107,5 +110,8 @@ export const accountAwardsFields = {
     cfdaTitle: 'award__transaction__assistance_data__cfda__program_title',
     naics: 'award__transaction__contract_data__naics',
     naicsDescription: 'award__transaction__contract_data__naics_description',
-    psc: 'award__transaction__contract_data__product_or_service_code'
+    psc: 'award__transaction__contract_data__product_or_service_code',
+    pricingType: 'award__latest_transaction__contract_data__type_of_contract_pricing',
+    setAside: 'award__latest_transaction__contract_data__type_set_aside',
+    extentCompeted: 'award__latest_transaction__contract_data__extent_competed'
```
Comment (russellbodine) [inline src/js/dataMapping/search/filterFields.js]: i believe these are messed up (my fault).  Change from latest_trasaction to transaction
Should be: 'award__transaction__contract_data__type_set_aside'
not: 'award__latest_transaction__contract_data__type_set_aside'

**Comment by kevinli-work** [comment]
Can you update this PR to pass tests?
Also, I thought we were doing an "Other Award Items" section instead of standalone sections for certain filters?
screen shot 2017-09-12 at 5 14 42 pm

**Comment by ejgullo** [comment]
@kevinli-work Not sure what was up w/ Sidebar - saw the same as you when i went to look at the issue, but on restart, it works fine with no code change.  Could even find 'Other Award Items' in the code.  Odd.
Tests should be passing now.

File: src/js/models/search/queryBuilders/ContractFilterQuery.js [hunk 10]
```diff
@@ -0,0 +1,70 @@
+/**
```
Comment (kevinli-work) [inline src/js/models/search/queryBuilders/ContractFilterQuery.js]: What is this file being used for/by?

Comment (ejgullo) [reply src/js/models/search/queryBuilders/ContractFilterQuery.js]: @kevinli-work It isn't.  Missed deleting when I moved these query builders to the `OtherFiltersQuery` file.  Removing.

**Pull request merged**